File: src/main/java/dev/mv/engine/render/shared/batch/Batch.java
package dev.mv.engine.render.shared.batch;

import dev.mv.engine.render.shared.Window;
import dev.mv.engine.render.shared.shader.Shader;
import dev.mv.engine.render.shared.texture.Texture;
import org.lwjgl.BufferUtils;

import java.nio.FloatBuffer;
import java.nio.IntBuffer;
import java.util.Arrays;

public abstract class Batch {
    public static final int POSITION_SIZE = 3;
    public static final int ROTATION_SIZE = 1;
    public static final int ROTATION_ORIGIN_SIZE = 2;
    public static final int COLOR_SIZE = 4;
    public static final int UV_SIZE = 2;
    public static final int TEX_ID_SIZE = 1;
    public static final int USE_CAMERA_SIZE = 1;
    public static final int TRANSFORM_ROTATION_SIZE = 1;
    public static final int TRANSFORM_TRANSLATE_SIZE = 2;
    public static final int TRANSFORM_ORIGIN_SIZE = 2;
    public static final int IS_FONT_SIZE = 1;

    public static final int VERTEX_SIZE_FLOATS = POSITION_SIZE + ROTATION_SIZE + ROTATION_ORIGIN_SIZE + COLOR_SIZE + UV_SIZE + TEX_ID_SIZE + USE_CAMERA_SIZE + TRANSFORM_ROTATION_SIZE + TRANSFORM_TRANSLATE_SIZE + TRANSFORM_ORIGIN_SIZE + IS_FONT_SIZE;
    public static final int VERTEX_SIZE_BYTES = VERTEX_SIZE_FLOATS * Float.BYTES;

    public static final int POSITION_OFFSET = 0;
    public static final int POSITION_OFFSET_BYTES = POSITION_OFFSET * Float.BYTES;
    public static final int ROTATION_OFFSET = POSITION_SIZE;
    public static final int ROTATION_OFFSET_BYTES = ROTATION_OFFSET * Float.BYTES;
    public static final int ROTATION_ORIGIN_OFFSET = ROTATION_OFFSET + ROTATION_SIZE;
    public static final int ROTATION_ORIGIN_OFFSET_BYTES = ROTATION_ORIGIN_OFFSET * Float.BYTES;
    public static final int COLOR_OFFSET = ROTATION_ORIGIN_OFFSET + ROTATION_ORIGIN_SIZE;
    public static final int COLOR_OFFSET_BYTES = COLOR_OFFSET * Float.BYTES;
    public static final int UV_OFFSET = COLOR_OFFSET + COLOR_SIZE;
    public static final int UV_OFFSET_BYTES = UV_OFFSET * Float.BYTES;
    public static final int TEX_ID_OFFSET = UV_OFFSET + UV_SIZE;
    public static final int TEX_ID_OFFSET_BYTES = TEX_ID_OFFSET * Float.BYTES;
    public static final int USE_CAMERA_OFFSET = TEX_ID_OFFSET + TEX_ID_SIZE;
    public static final int USE_CAMERA_OFFSET_BYTES = USE_CAMERA_OFFSET * Float.BYTES;
    public static final int TRANSFORM_ROTATION_OFFSET = USE_CAMERA_OFFSET + USE_CAMERA_SIZE;
    public static final int TRANSFORM_ROTATION_OFFSET_BYTES = TRANSFORM_ROTATION_OFFSET * Float.BYTES;
    public static final int TRANSFORM_TRANSLATE_OFFSET = TRANSFORM_ROTATION_OFFSET + TRANSFORM_ROTATION_SIZE;
    public static final int TRANSFORM_TRANSLATE_OFFSET_BYTES = TRANSFORM_TRANSLATE_OFFSET * Float.BYTES;
    public static final int TRANSFORM_ORIGIN_OFFSET = TRANSFORM_TRANSLATE_OFFSET + TRANSFORM_TRANSLATE_SIZE;
    public static final int TRANSFORM_ORIGIN_OFFSET_BYTES = TRANSFORM_ORIGIN_OFFSET * Float.BYTES;
    public static final int IS_FONT_OFFSET = TRANSFORM_ORIGIN_OFFSET + TRANSFORM_ORIGIN_SIZE;
    public static final int IS_FONT_OFFSET_BYTES = IS_FONT_OFFSET * Float.BYTES;
    // p p p r ro ro c c c c uv uv ti uc tr tt tt to to
    protected int maxSize;
    protected float[] data;
    protected int[] indices;
    protected Texture[] textures;
    protected Window win;
    protected Shader shader;
    protected FloatBuffer vbo;
    protected int vbo_id;
    protected IntBuffer ibo;
    protected int ibo_id;
    protected int[] tex_ids;
    protected boolean isStencil = false;
    /**
     * The var vertCount is the offset pointer for the incoming data,
     * therefor no data gets overridden.
     * For clearing this var, use clearBatch().
     */

    protected int vertCount = 0;
    protected int objCount = 0;
    protected int nextFreeTexSlot = 0;
    protected boolean isFull = false;
    protected boolean isFullTex = false;

    public Batch(int maxSize, Window win, Shader shader, boolean isStencil) {
        this.maxSize = maxSize;
        this.win = win;
        this.shader = shader;
        this.isStencil = isStencil;
        initBatch();
    }

    public Batch(int maxSize, Window win, Shader shader) {
        this(maxSize, win, shader, false);
    }

    private void initBatch() {
        data = new float[VERTEX_SIZE_FLOATS * maxSize];
        indices = new int[maxSize * 6];

        vbo = BufferUtils.createFloatBuffer(VERTEX_SIZE_BYTES * maxSize);
        vbo_id = win.getRender().genBuffers();

        ibo = BufferUtils.createIntBuffer(maxSize * 6);
        ibo_id = win.getRender().genBuffers();

        textures = new Texture[17];
        tex_ids = new int[17];
    }

    /**
     * Important Note:
     * This method does not really clear the data of this batch, it just sets the data offset back to 0.
     * With this change, the batch gets overridden if new data comes in.
     * This is better for performance than actually clearing the array.
     * for really clearing the data, use forceClearBatch().
     */

    public void clearBatch() {
        vertCount = 0;
        objCount = 0;
        nextFreeTexSlot = 0;

        isFull = false;
        isFullTex = false;
    }

    /**
     * This method clears the actual data of the data array.
     * But keep in mind, that this uses some performance and for only resetting the data
     * offset, use clearBatch().
     */

    public void forceClearBatch() {
        Arrays.fill(data, 0, (vertCount * VERTEX_SIZE_FLOATS) + 1, 0);
        Arrays.fill(textures, 0, nextFreeTexSlot, null);
        Arrays.fill(tex_ids, 0, nextFreeTexSlot, 0);
        vertCount = 0;
        objCount = 0;
        nextFreeTexSlot = 0;

        isFull = false;
        isFullTex = false;
    }

    public boolean isFull(int amount) {
        return (vertCount * VERTEX_SIZE_FLOATS) + amount >= maxSize;
    }

    public boolean isFullOfTextures() {
        return isFullTex;
    }

    private void addVertex(Vertex vertex) {
        for (int i = 0; i < VERTEX_SIZE_FLOATS; i++) {
            data[i + (vertCount * VERTEX_SIZE_FLOATS)] = vertex.get(i);
        }
        vertCount++;
    }

    public void addVertices(VertexGroup vertData, boolean useCamera, float tR, int tTx, int tTy, int tOx, int tOy, boolean isFont) {
        if (isFull(vertData.length())) return;

        genIndices(vertData.length());

        for (int i = 0; i < vertData.length(); i++) {
            addVertex(vertData.get(i)
                    .add(useCamera ? 1 : 0)
                    .add(tR)
                    .add(tTx)
                    .add(tTy)
                    .add(tOx)
                    .add(tOy)
                    .add(isFont ? 1 : 0)
            );
            if (vertCount > maxSize) {
                isFull = true;
                return;
            }
        }
        if (vertData.length() < 4) {
            addVertex(vertData.get(0)
                    .add(useCamera ? 1 : 0)
                    .add(tR)
                    .add(tTx)
                    .add(tTy)
                    .add(tOx)
                    .add(tOy)
                    .add(isFont ? 1 : 0)
            );
            if (vertCount > maxSize) {
                isFull = true;
                return;
            }
        }

        objCount++;
    }

    public int addTexture(Texture tex) {

        if (isFullTex) return -1;

        for (int i = 0; i < textures.length; i++) {
            if (textures[i] == null) continue;
            if (textures[i].getId() == tex.getId()) {
                return i + 1;
            }
        }

        textures[nextFreeTexSlot] = tex;
        tex_ids[nextFreeTexSlot] = nextFreeTexSlot + 1;
        nextFreeTexSlot++;

        if (nextFreeTexSlot >= textures.length) isFullTex = true;

        return nextFreeTexSlot;
    }

    public void finish() {
        vbo.put(data);
        ibo.put(indices);
        vbo.flip();
        ibo.flip();
    }

    public void render() {
        win.getRender().retrieveVertexData(textures, tex_ids, indices, data, vbo_id, ibo_id, shader, win.getAdapter().adaptRenderMode(getRenderMode()), isStencil);

        forceClearBatch();
    }

    public Shader getShader() {
        return shader;
    }

    public void setShader(Shader shader) {
        this.shader = shader;
    }

    public boolean isStencil() {
        return isStencil;
    }

    public abstract int getRenderMode();

    protected abstract void genIndices(int vertAmount);

    public abstract boolean isStrip();
}

File: src/render/batch.rs
use std::rc::Rc;
use glfw::Key::N;
use mvutils::utils::{IncDec, TetrahedronOp};
use crate::render::shared::{RenderProcessor2D, Shader, Texture, TextureRegion, Window};
use std::cell::RefCell;
use std::ops::Deref;
use mvutils::init_arr;

pub(crate) const FLOAT_BYTES: u16 = 4;

pub mod batch_layout_2d {
    use crate::render::batch::FLOAT_BYTES;

    pub(crate) const POSITION_SIZE: u16 = 3;
    pub(crate) const ROTATION_SIZE: u16 = 1;
    pub(crate) const ROTATION_ORIGIN_SIZE: u16 = 2;
    pub(crate) const COLOR_SIZE: u16 = 4;
    pub(crate) const UV_SIZE: u16 = 2;
    pub(crate) const TEX_ID_SIZE: u16 = 1;
    pub(crate) const CANVAS_COORDS_SIZE: u16 = 4;
    pub(crate) const CANVAS_DATA_SIZE: u16 = 2;
    pub(crate) const USE_CAMERA_SIZE: u16 = 1;
    pub(crate) const VERTEX_SIZE_FLOATS: u16 = POSITION_SIZE + ROTATION_SIZE + ROTATION_ORIGIN_SIZE + COLOR_SIZE + UV_SIZE + TEX_ID_SIZE + CANVAS_COORDS_SIZE + CANVAS_DATA_SIZE + USE_CAMERA_SIZE;
    pub(crate) const VERTEX_SIZE_BYTES: u16 = VERTEX_SIZE_FLOATS * FLOAT_BYTES;
    pub(crate) const POSITION_OFFSET: u16 = 0;
    pub(crate) const POSITION_OFFSET_BYTES: u16 = POSITION_OFFSET * FLOAT_BYTES;
    pub(crate) const ROTATION_OFFSET: u16 = POSITION_SIZE;
    pub(crate) const ROTATION_OFFSET_BYTES: u16 = ROTATION_OFFSET * FLOAT_BYTES;
    pub(crate) const ROTATION_ORIGIN_OFFSET: u16 = ROTATION_OFFSET + ROTATION_SIZE;
    pub(crate) const ROTATION_ORIGIN_OFFSET_BYTES: u16 = ROTATION_ORIGIN_OFFSET * FLOAT_BYTES;
    pub(crate) const COLOR_OFFSET: u16 = ROTATION_ORIGIN_OFFSET + ROTATION_ORIGIN_SIZE;
    pub(crate) const COLOR_OFFSET_BYTES: u16 = COLOR_OFFSET * FLOAT_BYTES;
    pub(crate) const UV_OFFSET: u16 = COLOR_OFFSET + COLOR_SIZE;
    pub(crate) const UV_OFFSET_BYTES: u16 = UV_OFFSET * FLOAT_BYTES;
    pub(crate) const TEX_ID_OFFSET: u16 = UV_OFFSET + UV_SIZE;
    pub(crate) const TEX_ID_OFFSET_BYTES: u16 = TEX_ID_OFFSET * FLOAT_BYTES;
    pub(crate) const CANVAS_COORDS_OFFSET: u16 = TEX_ID_OFFSET + TEX_ID_SIZE;
    pub(crate) const CANVAS_COORDS_OFFSET_BYTES: u16 = CANVAS_COORDS_OFFSET * FLOAT_BYTES;
    pub(crate) const CANVAS_DATA_OFFSET: u16 = CANVAS_COORDS_OFFSET + CANVAS_COORDS_SIZE;
    pub(crate) const CANVAS_DATA_OFFSET_BYTES: u16 = CANVAS_DATA_OFFSET * FLOAT_BYTES;
    pub(crate) const USE_CAMERA_OFFSET: u16 = CANVAS_DATA_OFFSET + CANVAS_DATA_OFFSET_BYTES;
    pub(crate) const USE_CAMERA_OFFSET_BYTES: u16 = USE_CAMERA_OFFSET * FLOAT_BYTES;
}

pub(crate) trait BatchGen {
    fn get_render_mode(&self) -> u8;
    fn gen_indices(&self, amt: u16, offset: u32, indices: &mut Vec<u32>);
    fn is_stripped(&self) -> bool;
}

struct RegularBatch;
struct StrippedBatch;

impl BatchGen for RegularBatch {
    fn get_render_mode(&self) -> u8 {
        gl::TRIANGLES as u8
    }

    fn gen_indices(&self, amt: u16, offset: u32, indices: &mut Vec<u32>) {
        if amt == 4 {
            indices[(offset * 6 + 0) as usize] = 0 + offset * 4;
            indices[(offset * 6 + 1) as usize] = 1 + offset * 4;
            indices[(offset * 6 + 2) as usize] = 2 + offset * 4;
            indices[(offset * 6 + 3) as usize] = 0 + offset * 4;
            indices[(offset * 6 + 4) as usize] = 2 + offset * 4;
            indices[(offset * 6 + 5) as usize] = 3 + offset * 4;
        } else {
            indices.insert((offset * 6 + 0) as usize, 0 + offset * 4);
            indices.insert((offset * 6 + 1) as usize, 1 + offset * 4);
            indices.insert((offset * 6 + 2) as usize, 2 + offset * 4);
        }
    }

    fn is_stripped(&self) -> bool {
        false
    }
}

impl BatchGen for StrippedBatch {
    fn get_render_mode(&self) -> u8 {
        gl::TRIANGLE_STRIP as u8
    }

    fn gen_indices(&self, amt: u16, offset: u32, indices: &mut Vec<u32>) {
        if amt == 4 {
            indices[(offset * 6 + 0) as usize] = 0 + offset * 4;
            indices[(offset * 6 + 1) as usize] = 1 + offset * 4;
            indices[(offset * 6 + 2) as usize] = 2 + offset * 4;
            indices[(offset * 6 + 3) as usize] = 0 + offset * 4;
            indices[(offset * 6 + 4) as usize] = 2 + offset * 4;
            indices[(offset * 6 + 5) as usize] = 3 + offset * 4;
        } else {
            indices[(offset * 6 + 0) as usize] = 0 + offset * 4;
            indices[(offset * 6 + 1) as usize] = 1 + offset * 4;
            indices[(offset * 6 + 2) as usize] = 2 + offset * 4;
        }
    }

    fn is_stripped(&self) -> bool {
        true
    }
}

struct Batch2D {
    generator: Box<dyn BatchGen>,
    data: Vec<f32>,
    indices: Vec<u32>,
    textures: [Option<Rc<RefCell<Texture>>>; 17],
    tex_ids: [u32; 17],
    shader: Rc<RefCell<Shader>>,
    vbo: u32,
    ibo: u32,
    size: u32,
    vert_count: u32,
    obj_count: u32,
    next_tex: u32,
    full: bool,
    full_tex: bool
}

impl Batch2D {
    pub(crate) fn new<T: BatchGen + 'static>(size: u32, shader: Rc<RefCell<Shader>>, generator: T) -> Self {
        Batch2D {
            generator: Box::new(generator),
            data: Vec::with_capacity(size as usize * batch_layout_2d::VERTEX_SIZE_FLOATS as usize),
            indices: Vec::with_capacity(size as usize * 6),
            textures: [0; 17].map(|n| None),
            tex_ids: [0; 17],
            shader,
            vbo: 0,
            ibo: 0,
            size,
            vert_count: 0,
            obj_count: 0,
            next_tex: 0,
            full: false,
            full_tex: false
        }
    }

    pub(crate) fn clear(&mut self) {
        self.vert_count = 0;
        self.obj_count = 0;
        self.next_tex = 0;

        self.full = false;
        self.full_tex = false;
    }

    pub(crate) fn force_clear(&mut self) {
        self.data.clear();
        self.indices.clear();
        self.textures.fill(None);
        self.tex_ids.fill(0);

        self.clear();
    }

    pub(crate) fn is_full(&self, amount: u32) -> bool {
        self.data.capacity() < amount as usize
    }

    pub(crate) fn is_full_tex(&self) -> bool {
        self.full_tex
    }

    fn add_vertex(&mut self, vertex: &Vertex2D) {
        for i in 0..batch_layout_2d::VERTEX_SIZE_FLOATS {
            self.data.insert(i as usize + (self.vert_count * batch_layout_2d::VERTEX_SIZE_FLOATS as u32) as usize, vertex.data[i as usize]);
        }
        self.vert_count += 1;
    }

    pub(crate) fn add_vertices(&mut self, vertices: &VertexGroup<Vertex2D>) {
        if self.is_full(vertices.len as u32) {
            return;
        }

        self.generator.gen_indices(vertices.len as u16, self.obj_count, &mut self.indices);

        for i in 0..vertices.len {
            self.add_vertex(vertices.get(i));
            if self.vert_count > self.size {
                self.full = true;
                return;
            }
        }
        if vertices.len < 4 {
            self.add_vertex(vertices.get(0));
            if self.vert_count > self.size {
                self.full = true;
                return;
            }
        }

        self.obj_count += 1;
    }

    pub(crate) fn add_texture(&mut self, texture: Rc<RefCell<Texture>>) -> u32 {
        if self.full_tex {
            return 0;
        }

        for i in 0..17 {
            if let Some(tex) = &self.textures[i] {
                if tex.borrow().get_id() == texture.borrow().get_id() {
                    return i as u32 + 1;
                }
            }
        }

        self.textures[self.next_tex as usize] = Some(texture);
        self.tex_ids[self.next_tex as usize] = self.next_tex + 1;
        self.next_tex += 1;

        if self.next_tex >= 17 {
            self.full_tex = true;
        }

        return self.next_tex;
    }

    pub(crate) fn render(&mut self, processor: &impl RenderProcessor2D) {
        if self.vbo == 0 {
            self.vbo = processor.gen_buffer_id();
        }
        if self.ibo == 0 {
            self.ibo = processor.gen_buffer_id();
        }
        processor.process_data(&mut self.textures, &self.tex_ids, &self.indices, &self.data, self.vbo, self.ibo, self.shader.borrow().deref(), processor.adapt_render_mode(self.generator.get_render_mode()));
        self.force_clear();
    }

    pub(crate) fn set_shader(&mut self, shader: Rc<RefCell<Shader>>) {
        self.shader = shader;
    }

    pub(crate) fn is_stripped(&self) -> bool {
        self.generator.is_stripped()
    }
}

//data storage

pub(crate) struct Vertex2D {
    data: [f32; batch_layout_2d::VERTEX_SIZE_FLOATS as usize]
}

impl Vertex2D {
    pub(crate) fn new() -> Self {
        Vertex2D { data: [0.0; batch_layout_2d::VERTEX_SIZE_FLOATS as usize] }
    }

    pub(crate) fn set(&mut self, data: [f32; batch_layout_2d::VERTEX_SIZE_FLOATS as usize]) {
        self.data = data;
    }
}

pub(crate) struct VertexGroup<T> {
    vertices: [T; 4],
    len: u8
}

impl VertexGroup<Vertex2D> {
    pub(crate) fn new() -> Self {
        VertexGroup {
            vertices: init_arr!(4, Vertex2D::new()),
            len: 0
        }
    }

    pub(crate) fn get(&self, index: u8) -> &Vertex2D {
        assert!(index <= 4);
        &self.vertices[index as usize]
    }

    pub(crate) fn get_mut(&mut self, index: u8) -> &mut Vertex2D {
        assert!(index <= 4);
        &mut self.vertices[index as usize]
    }

    pub(crate) fn set_len(&mut self, len: u8) {
        self.len = len;
    }

    pub(crate) fn len(&self) -> u8 {
        self.len
    }
}

//controllers

pub(crate) struct BatchController2D {
    batches: Vec<Batch2D>,
    batch_limit: u32,
    shader: Rc<RefCell<Shader>>,
    default_shader: Rc<RefCell<Shader>>,
    current: u32
}

impl BatchController2D {
    pub(crate) fn new(shader: Rc<RefCell<Shader>>, batch_limit: u32) -> Self {
        assert!(batch_limit >= 14);
        shader.borrow_mut().make();
        shader.borrow_mut().bind();
        let mut batch = BatchController2D {
            batches: Vec::new(),
            batch_limit,
            default_shader: shader.clone(),
            shader,
            current: 0
        };
        batch.start();
        batch
    }

    pub(crate) fn start(&mut self) {
        self.batches.push(Batch2D::new(self.batch_limit, self.shader.clone(), RegularBatch));
    }

    fn next_batch(&mut self, stripped: bool) {
        self.current += 1;
        if let Some(batch) = self.batches.get(self.current as usize) {
            if batch.is_stripped() != stripped {
                self.batches.insert(self.current as usize, self.gen_batch(stripped))
            }
        } else {
            self.batches.push(self.gen_batch(stripped));
        }

    }

    fn gen_batch(&self, stripped: bool) -> Batch2D {
        stripped.yn(Batch2D::new(self.batch_limit, self.shader.clone(), StrippedBatch),
                    Batch2D::new(self.batch_limit, self.shader.clone(), RegularBatch))
    }

    pub(crate) fn add_vertices(&mut self, vertices: &VertexGroup<Vertex2D>) {
        if self.batches[self.current as usize].is_stripped() {
            self.next_batch(false);
        }
        if self.batches[self.current as usize ].is_full(vertices.len as u32 * batch_layout_2d::VERTEX_SIZE_FLOATS as u32) {
            self.next_batch(false);
        }

        self.batches[self.current as usize].add_vertices(vertices);
    }

    pub(crate) fn add_vertices_stripped(&mut self, vertices: &VertexGroup<Vertex2D>) {
        if !self.batches[self.current as usize].is_stripped() {
            self.next_batch(true);
        }
        if self.batches[self.current as usize].is_full(vertices.len as u32 * batch_layout_2d::VERTEX_SIZE_FLOATS as u32) {
            self.next_batch(true);
        }

        self.batches[self.current as usize].add_vertices(vertices);
    }

    pub(crate) fn add_texture(&mut self, texture: Rc<RefCell<Texture>>, vert_count: u32) -> u32 {
        if self.batches[self.current as usize].is_stripped() {
            self.next_batch(false);
        }
        if self.batches[self.current as usize].is_full(vert_count * batch_layout_2d::VERTEX_SIZE_FLOATS as u32) {
            self.next_batch(false);
        }

        self.batches[self.current as usize].add_texture(texture)
    }

    pub(crate) fn add_texture_stripped(&mut self, texture: Rc<RefCell<Texture>>, vert_count: u32) -> u32 {
        if !self.batches[self.current as usize].is_stripped() {
            self.next_batch(true);
        }
        if self.batches[self.current as usize].is_full(vert_count * batch_layout_2d::VERTEX_SIZE_FLOATS as u32) {
            self.next_batch(true);
        }

        self.batches[self.current as usize].add_texture(texture)
    }

    pub(crate) fn render(&mut self, processor: &impl RenderProcessor2D) {
        for i in 0..self.current + 1 {
            self.batches[i as usize].render(processor);
        }
        self.current = 0;
    }

    pub(crate) fn set_shader(&mut self, shader: Rc<RefCell<Shader>>) {
        self.shader = shader;
        self.shader.borrow_mut().make();
        for batch in self.batches.iter_mut() {
            batch.set_shader(self.shader.clone());
        }
    }

    pub(crate) fn reset_shader(&mut self) {
        self.set_shader(self.default_shader.clone());
    }
}